hathor/consensus/poa/poa_block_producer.py

```diff
@@ -0,0 +1,172 @@
+        self._schedule_block_lc.clock = self._reactor
+
+    def start(self) -> None:
+        self._start_producing_lc.start(_WAIT_SYNC_DELAY)
+        self._schedule_block_lc.start(self._settings.AVG_TIME_BETWEEN_BLOCKS)
+
+    def stop(self) -> None:
+        if self._start_producing_lc.running:
+            self._start_producing_lc.stop()
+
+        if self._schedule_block_lc.running:
+            self._schedule_block_lc.stop()
+
+    @staticmethod
+    def _calculate_signer_index(settings: PoaSettings, poa_signer: PoaSigner) -> int:
+        """Return the signer index for the given private key."""
+        public_key = poa_signer.get_public_key()
+        public_key_bytes = get_public_key_bytes_compressed(public_key)
+        sorted_signers = sorted(settings.signers)
+        try:
+            return sorted_signers.index(public_key_bytes)
+        except ValueError:
+            raise ValueError(f'Public key "{public_key_bytes.hex()}" not in list of PoA signers')
+
+    def _start_producing(self) -> None:
+        """Start producing new blocks."""
+        if not self._manager.can_start_mining():
+            # We're syncing, so we'll try again later
+            self._log.warn('cannot start producing new blocks, node not synced')
+            return
+
+        self._log.info('started producing new blocks')
+        self._started_producing = True
+        self._start_producing_lc.stop()
+
+    def _schedule_block(self) -> None:
+        """Schedule propagation of a new block."""
+        previous_block = self._manager.tx_storage.get_best_block()
+        assert isinstance(previous_block, PoaBlock)
+        if not self._started_producing or previous_block == self._last_seen_best_block:
+            return
+
+        self._last_seen_best_block = previous_block
+        now = self._reactor.seconds()
+        expected_timestamp = self._expected_block_timestamp(previous_block)
+        propagation_delay = 0 if expected_timestamp < now else expected_timestamp - now
+
+        self._reactor.callLater(propagation_delay, self._produce_block, previous_block)
+        self._log.debug(
+            'scheduling block production',
+            previous_block=previous_block.hash_hex,
+            previous_block_height=previous_block.get_height(),
+            delay=propagation_delay,
+        )
+
+    def _produce_block(self, previous_block: PoaBlock) -> None:
+        """Create and propagate a new block."""
+        block_templates = self._manager.get_block_templates(parent_block_hash=previous_block.hash)
+        block = block_templates.generate_mining_block(self._manager.rng, poa_block=True)
+        assert isinstance(block, PoaBlock)
+        block.weight = poa.calculate_weight(self._poa_settings, block, self._signer_index)
+        self._poa_signer.sign_block(block)
+        block.update_hash()
+
+        self._manager.on_new_tx(block, propagate_to_peers=False, fails_silently=False)
+        if not block.get_metadata().voided_by:
```

Comment:
@msbrogli commented on https://github.com/HathorNetwork/hathor-core-side-dag/pull/11#discussion_r1634980005:

> Nice rule. I guess we can add it to the `on_new_tx()` propagation part.

Reply:
@msbrogli maybe, but I would rather not change it on this PR, as it may affect PoW block propagation on mainnet so it deserves more thought. For instance, could it be possible that a node observes a block as voided, propagates it, an then other nodes (which may be synced to different best chains), observe it as non-voided?

Reply:
I agree to do it in a separate PR. Let's discuss it there. As a quick answer: I guess your situation would be fine because the sync algorithm would get these peers sync'ed anyway.

Reply:
Created issue https://github.com/HathorNetwork/hathor-core/issues/1066

Reply:
Thanks!